src/ssh-generator/ssh-generator.c

```diff
@@ -0,0 +1,341 @@
+ *     3. Listen on AF_UNIX socket /run/ssh-unix-local/socket (always)
+ *
+ * The first two provide a nice way for hosts to connect to containers and VMs they invoke via the usual SSH
+ * logic, but without waiting for networking or suchlike. The latter allows the same for local clients.
+ *
+ * Note: this does *not* make SSH listen on IP. That's left for users/packagers to enable manually. */
+
+// fixme: also by cred and per kernel cmdline
+// fixme: log command line to use for connection explicitly
+// fixme: also do machine/ which used machined supplied info and automatically connects to the right place
+
+static const char *arg_dest = NULL;
+static bool arg_enabled = true;
+
+static int parse_proc_cmdline_item(const char *key, const char *value, void *data) {
+        int r;
+
+        assert(key);
+
+        if (proc_cmdline_key_streq(key, "systemd.ssh_auto")) {
+                r = value ? parse_boolean(value) : 1;
+                if (r < 0)
+                        log_warning_errno(r, "Failed to parse systemd.ssh_auto switch \"%s\", ignoring: %m", value);
+                else
+                        arg_enabled = r;
+        }
+
+        return 0;
+}
+
+static int make_sshd_template_unit(
+                const char *dest,
+                const char *template,
+                const char *sshd_binary,
+                const char *found_sshd_template_service,
+                char **generated_sshd_template_unit) {
+
+        int r;
+
+        assert(dest);
+        assert(template);
+        assert(sshd_binary);
+        assert(generated_sshd_template_unit);
+
+        /* If the system has a suitable template already, symlink it to the name we want to reuse it */
+        if (found_sshd_template_service)
+                return generator_add_symlink(
+                                dest,
+                                template,
+                                /* dep_type= */ NULL,
+                                found_sshd_template_service);
+
+        if (!*generated_sshd_template_unit) {
+                _cleanup_fclose_ FILE *f = NULL;
+
+                r = generator_open_unit_file_full(
+                                dest,
+                                /* source= */ NULL,
+                                "sshd-generated@.service", /* Give this generated unit a generic name, since we want to use it for both AF_UNIX and AF_VSOCK */
+                                &f,
+                                generated_sshd_template_unit,
+                                /* ret_temp_path= */ NULL);
+                if (r < 0)
+                        return r;
+
+                fprintf(f,
+                        "[Unit]\n"
+                        "Description=OpenSSH Per-Connection Server Daemon\n"
+                        "Documentation=man:systemd-ssh-generator(8) man:sshd(8)\n"
+                        "[Service]\n"
+                        "ExecStart=-%s\n"
```

Comment:
Do we need `-` here since we explicitly look up `sshd` beforehand?

Also, isn't this missing `-i` to tell sshd to work off standard input/standard output?

Reply:
> Do we need `-` here since we explicitly look up `sshd` beforehand?

Yeah, sshd returns non-zero exit codes in some cases, it's really weird. I think it depends on the read error they get on the connection socket they sometimes turn that into exit 0 and sometimes something else. Hence we really should ignore the exit code here, otherwise we'll collect failed per-connection services.